physical/azure/azure_ha.go

```diff
@@ -0,0 +1,422 @@
+		case <-l.stopCh:
+			break OUTER
+		}
+
+		// Attempt to read the key
+		r, err := l.get(context.Background())
+		if err != nil {
+			retries++
+			continue
+		}
+
+		// Verify the identity is the same
+		if r == nil || r.Identity != l.identity {
+			break OUTER
+		}
+	}
+
+	l.stopLock.Lock()
+	defer l.stopLock.Unlock()
+	if !l.stopped {
+		l.stopped = true
+		close(l.stopCh)
+	}
+}
+
+// writeLock writes the given lock using the following algorithm:
+//
+// - lock does not exist
+//   - write the lock
+// - lock exists
+//   - if key is empty or identity is the same or timestamp exceeds TTL
+//     - update the lock to self
+func (l *Lock) writeLock() (bool, error) {
+	l.backend.permitPool.Acquire()
+	defer l.backend.permitPool.Release()
+
+	// Create a transaction to read and the update (maybe)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// The operation may be retried, so we need to stop it if we lose leadership.
+	go func() {
+		select {
+		case <-l.stopCh:
+			cancel()
+		case <-ctx.Done():
+		}
+	}()
+
+	// Build up the access conditions
+	var conds azblob.BlobAccessConditions
+
+	// Read the record
+	r, err := l.get(ctx)
+	if err != nil {
+		return false, errwrap.Wrapf("write lock: {{err}}", err)
+	}
+	if r != nil {
+		// If the key is empty or the identity is ours or the ttl expired, we can
+		// write. Otherwise, return now because we cannot.
+		if r.Key != "" && r.Identity != l.identity && time.Now().UTC().Sub(r.Timestamp) < l.ttl {
+			return false, nil
+		}
+		conds.IfMatch = r.etag // CAS mechanism
+	} else {
+		// IfNoneMatch: "*" = perform the operation iff the resource does not exist
+		// Avoid overwriting a lock created by someone else
+		conds.IfNoneMatch = "*"
+	}
+
+	// Update the lock to now
+	lockData, err := json.Marshal(&LockRecord{
+		Key:       l.key,
+		Value:     l.value,
+		Identity:  l.identity,
+		Timestamp: time.Now().UTC(),
+	})
+	if err != nil {
+		return false, errwrap.Wrapf("write lock: failed to encode JSON: {{err}}", err)
+	}
+
+	// Write the object
+	blob := l.backend.container.NewBlockBlobURL(l.key)
+
+	headers := azblob.BlobHTTPHeaders{
+		CacheControl: "no-cache; no-store; max-age=0",
+	}
+	metadata := azblob.Metadata(map[string]string{
+		"lock": string(lockData),
+	})
+
+	if resp, err := blob.Upload(context.Background(), nil, headers, metadata, conds, azblob.DefaultAccessTier, nil); err != nil {
+		if resp != nil && resp.StatusCode() == http.StatusPreconditionFailed {
+			return false, nil
+		}
+		return false, err
+	}
+
+	return true, nil
+}
+
+// get retrieves the value for the lock.
+func (l *Lock) get(ctx context.Context) (*LockRecord, error) {
+	l.backend.permitPool.Acquire()
+	defer l.backend.permitPool.Release()
+
+	// Check blob existence
+	blob := l.backend.container.NewBlobURL(l.key)
+	bProp, err := blob.GetProperties(context.Background(), azblob.BlobAccessConditions{})
+	if err != nil {
+		if bProp != nil && bProp.StatusCode() == http.StatusNotFound {
```

Comment:
I did some testing with this PR and had some trouble at bootstrapping. The absence of a lock is treated as error: The vault instances always yielded messages like `lock: attempt lock: write lock: failed to read blob properties for "core/lock":`, followed by a 404 error.

It seems that this line here is intended to treat 404s differently than other errors, as these 404s are part of normal operation, as one Vault instance has to create the lock initially (I imagine).

When doing a little bit of debugging, I found that `bProp` was `nil` also for 404s though, hence not satisfying this `if` clause, causing effectively a deadlock with aforementioned log output.

Unfortunately, as of now, I cannot provide a solution, though.

Reply:
fixed in bb3a3e2